astroquery/eso/utils.py

```diff
@@ -0,0 +1,197 @@
+from typing import Dict, List, Optional, Union
+from astropy.table import Table
+
+DEFAULT_LEAD_COLS_RAW = ['object', 'ra', 'dec', 'dp_id', 'date_obs', 'prog_id']
+DEFAULT_LEAD_COLS_PHASE3 = ['target_name', 's_ra', 's_dec', 'dp_id', 'date_obs', 'proposal_id']
+
+
+@dataclass
+class _UserParams:
+    """
+    Parameters set by the user
+    """
+    table_name: str
+    column_name: str = None
+    allowed_values: Union[List[str], str] = None
+    cone_ra: float = None
+    cone_dec: float = None
+    cone_radius: float = None
+    columns: Union[List, str] = None
+    column_filters: Dict[str, str] = None
+    top: int = None
+    order_by: str = ''
+    order_by_desc: bool = True
+    count_only: bool = False
+    query_str_only: bool = False
+    print_help: bool = False
+    authenticated: bool = False
+
+
+def _split_str_as_list_of_str(column_str: str):
+    if column_str == '':
+        column_list = []
+    else:
+        column_list = list(map(lambda x: x.strip(), column_str.split(',')))
+    return column_list
+
+
+def _raise_if_has_deprecated_keys(filters: Optional[Dict[str, str]]) -> bool:
+    if not filters:
+        return
+
+    if any(k in filters for k in ("box", "coord1", "coord2")):
+        raise ValueError(
+            "box, coord1 and coord2 are deprecated; "
+            "use cone_ra, cone_dec and cone_radius instead."
+        )
+
+    if any(k in filters for k in ("etime", "stime")):
+        raise ValueError(
+            "'stime' and 'etime' are deprecated; "
+            "use instead 'exp_start' together with '<', '>', 'between'. Examples:\n"
+            "\tcolumn_filters = {'exp_start': '< 2024-01-01'}\n"
+            "\tcolumn_filters = {'exp_start': '>= 2023-01-01'}\n"
+            "\tcolumn_filters = {'exp_start': \"between '2023-01-01' and '2024-01-01'\"}\n"
+        )
+
+
+def _build_where_constraints(
+        column_name: str,
+        allowed_values: Union[List[str], str],
+        column_filters: Dict[str, str]) -> str:
+    def _format_helper(av):
+        if isinstance(av, str):
+            av = _split_str_as_list_of_str(av)
+        quoted_values = [f"'{v.strip()}'" for v in av]
+        return f"{column_name} in ({', '.join(quoted_values)})"
+
+    column_filters = column_filters or {}
+    where_constraints = []
+    if allowed_values:
+        where_constraints.append(_format_helper(allowed_values))
+
+    where_constraints += [
+        f"{k} {_adql_sanitize_op_val(v)}" for k, v in column_filters.items()
+    ]
+    return where_constraints
+
+
+def reorder_columns(table: Table,
+                    leading_cols: Optional[List[str]] = None):
+    """
+    Reorders the columns of the pased table so that the
+    colums given by the list leading_cols are first.
+    If no leading cols are passed, it defaults to
+    ['object', 'ra', 'dec', 'dp_id', 'date_obs']
+    Returns a table with the columns reordered.
+    """
+    if not isinstance(table, Table):
+        return table
```

Comment:
Is this an expected case?

Reply:
yes, `table` may be a string or an int, depending on the user parameters `query_str_only` only and `count_only`.

I admit this is not the best software design at this point, but a complete refactoring would have taken more time from both developers and reviewers. We will come with better encapsulation, separation of concerns,  etc in the next pull request. Our aim for now was to switch to TAP with as little impact as possible to the API and to the overall code.